src/client/particles.cpp

```diff
@@ -31,635 +29,801 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 }
 
-void Particle::OnRegisterSceneNode()
+class CameraOffsetAffector : public irr::scene::IParticleAffector
 {
-	if (IsVisible)
-		SceneManager->registerNodeForRendering(this, scene::ESNRP_TRANSPARENT_EFFECT);
+public:
+	CameraOffsetAffector(const ClientEnvironment &env,
+		irr::scene::IParticleSystemSceneNode *ps, const v3f &position) :
+		env(env), ps(ps), position(position)
+	{
+	}
 
-	ISceneNode::OnRegisterSceneNode();
-}
+	void affect(u32 now, irr::scene::SParticle *particlearray, u32 count) override
+	{
+		v3s16 camera_offset = env.getCameraOffset();
+		ps->setPosition(position - intToFloat(camera_offset, BS));
+	}
 
-void Particle::render()
-{
-	video::IVideoDriver* driver = SceneManager->getVideoDriver();
-	driver->setMaterial(m_material);
-	driver->setTransform(video::ETS_WORLD, AbsoluteTransformation);
-
-	u16 indices[] = {0,1,2, 2,3,0};
-	driver->drawVertexPrimitiveList(m_vertices, 4,
-			indices, 2, video::EVT_STANDARD,
-			scene::EPT_TRIANGLES, video::EIT_16BIT);
-}
+	irr::scene::E_PARTICLE_AFFECTOR_TYPE getType() const override
+	{
+		return irr::scene::EPAT_NONE;
+	}
+private:
+	const ClientEnvironment &env;
+	irr::scene::IParticleSystemSceneNode *const ps;
+	v3f position;
+};
 
-void Particle::step(float dtime)
+class LightingAffector : public irr::scene::IParticleAffector
 {
-	m_time += dtime;
-	if (m_collisiondetection) {
-		aabb3f box = m_collisionbox;
-		v3f p_pos = m_pos * BS;
-		v3f p_velocity = m_velocity * BS;
-		collisionMoveResult r = collisionMoveSimple(m_env, m_gamedef, BS * 0.5f,
-			box, 0.0f, dtime, &p_pos, &p_velocity, m_acceleration * BS, nullptr,
-			m_object_collision);
-		if (m_collision_removal && r.collides) {
-			// force expiration of the particle
-			m_expiration = -1.0;
-		} else {
-			m_pos = p_pos / BS;
-			m_velocity = p_velocity / BS;
-		}
-	} else {
-		m_velocity += m_acceleration * dtime;
-		m_pos += m_velocity * dtime;
+public:
+	LightingAffector(ClientEnvironment *env, IGameDef *gamedef,
+		u8 glow, const video::SColor &base_color = video::SColor(0xFFFFFFFF))
+		: env(env), gamedef(gamedef), glow(glow),
+		base_color(base_color)
+	{
 	}
-	if (m_animation.type != TAT_NONE) {
-		m_animation_time += dtime;
-		int frame_length_i, frame_count;
-		m_animation.determineParams(
-				m_material.getTexture(0)->getSize(),
-				&frame_count, &frame_length_i, NULL);
-		float frame_length = frame_length_i / 1000.0;
-		while (m_animation_time > frame_length) {
-			m_animation_frame++;
-			m_animation_time -= frame_length;
+
+	void affect(u32 now, irr::scene::SParticle *particlearray, u32 count) override
+	{
+		for (u32 i = 0; i < count; ++i) {
+			v3f pos = particlearray[i].pos +
+				intToFloat(env->getCameraOffset(), BS);
+			color = getParticleLightColor(pos / BS);
+			particlearray[i].color = color;
 		}
 	}
 
-	// Update lighting
-	updateLight();
+	irr::scene::E_PARTICLE_AFFECTOR_TYPE getType() const override
+	{
+		return irr::scene::EPAT_NONE;
+	}
+private:
+	ClientEnvironment *env;
```

Comment:
Private members ought to have the prefix `m_`

Reply:
> ...or m_ for members are allowed.

Allowed, not required. https://dev.minetest.net/Code_style_guidelines

Reply:
If it helps readability, then why not? :)